lib/new_relic/agent/transaction/abstract_segment.rb

```diff
@@ -327,6 +331,55 @@ def transaction_state
             Tracer.state
           end
         end
+
+        # for segment callback usage info, see self.set_segment_callback
+        def invoke_callback
+          return unless self.class.instance_variable_defined?(CALLBACK)
+
+          NewRelic::Agent.logger.debug("Invoking callback for #{self.class.name}...")
+          self.class.instance_variable_get(CALLBACK).call
+        end
+
+        # Setting and invoking a segment callback
+        # =======================================
+        # Each individual segment class such as `ExternalRequestSegment` allows
+        # for exactly one instance of a `Proc` (meaning a proc or lambda) to be
+        # set as a callback. A callback can be set on a segment class by calling
+        # `.set_segment_callback` with a proc or lambda as the only argument.
+        # If set, the callback will be invoked with `#call` at segment class
+        # initialization time.
+        #
+        # Example usage:
+        #   callback = -> { puts 'Hello, World! }
+        #   ExternalRequestSegment.set_segment_callback(callback)
+        #   ExternalRequestSegment.new(library, uri, procedure)
+        #
+        # A callback set on a segment class will only be called when that
+        # specific segment class is initialized. Other segment classes will not
+        # be impacted.
+        #
+        # Great caution should be taken in the defining of the callback block
+        # to not have the block perform anything too time consuming or resource
+        # intensive in order to keep the New Relic Ruby agent operating
+        # normally.
+        #
+        # Given that callbacks are user defined, they must be set entirely at
+        # the user's own risk. It is recommended that each callback use
+        # conditional logic that only performs work for certain qualified
+        # segments. It is recommended that each callback be thoroughly tested
+        # in non-production environments before being introduced to production
+        # environments.
+        def self.set_segment_callback(callback_proc)
+          unless callback_proc.is_a?(Proc)
+            NewRelic::Agent.logger.error("#{self}.#{__method__}: expected an argument of type Proc, " \
+                                         "got #{callback_proc.class}")
+            return
+          end
+
+          label = "set_callback_#{name.split('::').last.downcase.sub(SEGMENT, NewRelic::EMPTY_STR)}".to_sym
+          NewRelic::Agent.record_api_supportability_metric(label)
```

Comment:
I'm worried this supportability metric will not get recorded. Looking at the code for `record_api_supportability_metric`, the method name seems like it needs to be added to  `NewRelic::SupportabilityHelper::API_SUPPORTABILITY_METRICS`.

https://github.com/newrelic/newrelic-ruby-agent/blob/62ee0257b81f229dc2c03c7c1bf564edec6b8ee8/lib/new_relic/supportability_helper.rb#L64-L71

Reply:
You're absolutely right. I'm trying to create a dynamic metric name based on the segment class the callback was set on.

Preference for one of these?

- Use a single "set_segment_callback" metric name
- Define metric names for every single segment class and have each class use their own

Reply:
Discussed offline. Addressed with [40790eb](https://github.com/newrelic/newrelic-ruby-agent/pull/2178/commits/40790ebaf40da4bd28f65c743f26bf10438bb4ad)